webhooks/ratio_validator.go

```diff
@@ -0,0 +1,112 @@
+package webhooks
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+
+// RatioValidator checks for every action in a namespace whether the Memory to CPU ratio limit is exceeded and will return a warning if it is.
+type RatioValidator struct {
+	client  client.Client
+	decoder *admission.Decoder
+
+	RatioLimit *resource.Quantity
+}
+
+// Handle handles the admission requests
+func (v *RatioValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if strings.HasPrefix(req.UserInfo.Username, "system:") {
+		// Is service account or kube system user: https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-subjects
+		return admission.Allowed("system user")
+	}
+
+	l := log.FromContext(ctx).WithName("webhook.validate-request-ratio.appuio.io")
+	l.V(3).WithValues("kind", req.Kind.Kind, "namespace", req.Namespace).Info("handling request")
+
+	r, err := v.getRatio(ctx, req.Namespace)
+	if err != nil {
+		return errored(http.StatusInternalServerError, err)
+	}
+
+	// If we are creating an object with resource requests, we add them to the current ratio
+	// We cannot easily do this when updating resources.
+	if req.Operation == admissionv1.Create {
+		switch req.Kind.Kind {
+		case "Pod":
+			pod := corev1.Pod{}
+			if err := v.decoder.Decode(req, &pod); err != nil {
+				return errored(http.StatusBadRequest, err)
+			}
+			r = r.RecordPod(pod)
+		case "Deployment":
+			deploy := appsv1.Deployment{}
+			if err := v.decoder.Decode(req, &deploy); err != nil {
+				return errored(http.StatusBadRequest, err)
+			}
+			r = r.RecordDeployment(deploy)
+		case "StatefulSet":
+			sts := appsv1.StatefulSet{}
+			if err := v.decoder.Decode(req, &sts); err != nil {
+				return errored(http.StatusBadRequest, err)
+			}
+			r = r.RecordStatefulSet(sts)
+		}
+	}
+
+	if r.Below(*v.RatioLimit) {
+		return admission.Response{
+			AdmissionResponse: admissionv1.AdmissionResponse{
+				Allowed:  true,
+				Warnings: []string{fmt.Sprintf("Current memory to CPU ratio of %s/core in this namespace is below the fair use ratio of %s/core\n This might lead to additional cost.", r, v.RatioLimit)},
+			}}
+	}
+	return admission.Allowed("ok")
+}
+
+func (v *RatioValidator) getRatio(ctx context.Context, ns string) (*Ratio, error) {
+	r := NewRatio()
+	pods := corev1.PodList{}
+	err := v.client.List(ctx, &pods, client.InNamespace(ns))
```

Comment:
This means one request per created pod right? The default client QPS of five could be a tiny bit too small. Maybe it's worth experimenting with [cache.NewInformer](https://pkg.go.dev/k8s.io/client-go/tools/cache#NewInformer) for some minutes?

Reply:
Actually no. I was surprised as well. The `controller-runtime` out of the box comes with informers for basically everything. So reading from a `controller-runtime` client basically never actually generates a request. :)